File: factorial/factorial.py
def factorial(n):
    """
    factorial calculates the factorial of a number
    take the number as a parameter of the function
    if the number is less than or equal to 1, return 1
    initialize the result to 1
    iterate from 2 to n+1
    multiply the result by the current number
    return the result
    time comeplexity: 0(n)
    space complexity: 0(1)
    """
    total = 1
    if n <= 1:
        return 1

    for i in range(2, n+1):
        total *= i
    return total


print(factorial(5))
print(factorial(6))
print(factorial(7))


def factorialRecursive(n):
    """_summary_

    Args:
        n (_type_): _description_

    Returns:
        _type_: _description_
        The factorial of n. If n is less than or equal to 1, it returns 1. 
        Otherwise, it returns the product of n and the factorial of (n-1).
    Complexity:
    Time: O(n)
    Space: O(n)

    Steps:
    1. Base case: if n is less than or equal to 1, return 1.
    2. Otherwise, return n times the factorial of (n-1).
    3. The function calls itself recursively until it reaches the base case.
    4. The function returns the product of n and the factorial of (n-1).
    5. The recursive calls are stored in the call stack until the base case is reached.
    6. The function returns the final result from the call stack.
    """

    if n <= 1:
        return 1

    return n * factorialRecursive(n-1)


print(factorial(5))

File: factorial/factorial.js
// function that calculates the factorail of a number
// create a function and pass the parameeter n
// create a variable total to store the factorial value and intialize it with 1
// if n is less thatn or equal to 1, return 1
// create a
// loop through the numbers from 2 to n and multiply it by the factorial
// return the factorial
// time complexity is O(n)
// space complexity is O(1)

const factorial = function (n) {
  let total = 1;
  if (n <= 1) return 1;
  for (let i = 2; i <= n; i++) {
    total *= i;
  }
  return total;
};

console.log(factorial(5)); // 120
console.log(factorial(4)); // 24
console.log(factorial(3)); // 6
console.log(factorial(2)); // 2

// function that calculates the factorial of a number using recursion
// create a function and pass the parameter n
// if n is less than or equal to 1, return 1
// return n * the factorial of n - 1
// time complexity is O(n)
// space complexity is O(n)

const factorialRecursive = function (n) {
  if (n <= 1) return 1;
  return n * factorialRecursive(n - 1);
};

console.log(factorialRecursive(5)); // 120
console.log(factorialRecursive(4)); // 24
console.log(factorialRecursive(3)); // 6
